### PR TITLE
fix: disable inline optimization by default

### LINKED ISSUE #264: Turn off inlining in optimization until UglifyJS issue is fixed.
This minimization bug has been open for almost two months: https://github.com/mishoo/UglifyJS2/issues/2842.  It has been referenced here elsewhere.  The bug occurs when UglifyJS attempts to inline a call to a function with a variable of the same name as a variable in the calling function.  Symptoms usually include "is not defined" or "assignment to constant variable".

In my experience, it's both easy to fall into and extremely time-consuming to diagnose.  

**I'd suggest turning `inline: false` in Webpack 4 until this is fixed.**  From the comments, it looks like UglifyJS2 is suffering from a lack of contributors with the skills and bandwidth needed to deal with things like this.

All of the Webpack 4 blogging has been about "Just let us do it for you."  IMHO, to fulfill the implied promise, I think you need to be reactive to component issues like this.

Related: `optimization.minimize` and `optimization.minimizer` (along with the rest of `optimization`) need documenting.  

For this particular issue, I'm using:
```
{
optimization:
            minimizer: [
               
                new UglifyJsPlugin({
                    parallel: true,  // Webpack default
                    cache: true      // Webpack default
                    uglifyOptions: {
                        /*
                            inlining is broken sometimes where inlined function uses the same variable name as inlining function.
                            See https://github.com/mishoo/UglifyJS2/issues/2842, https://github.com/mishoo/UglifyJS2/issues/2843
                         */
                        compress: { inline:false },
                    },
                })
            ],
}

```

---

This issue was moved from webpack/webpack#6798 by @evilebottnawi. Original issue was by @estaub.

### LINKED ISSUE #307: Webpack generated broken code in production mode
# Bug report

**What is the current behavior?**

Webpack generated broken code in production mode.

- The source code: [src/index.js](https://github.com/mysticatea/webpack-issue/blob/77dee08f91c4dabfdca0e7f9229148d688db63e6/src/index.js)
- The result code: [dist/main.js](https://github.com/mysticatea/webpack-issue/blob/77dee08f91c4dabfdca0e7f9229148d688db63e6/dist/main.js)

The result code contains a re-assignment to a `const` variable. It throws a fatal error in strict mode.

```js
const r=e[t];(r=r).fatal||r.ruleId||!r.message.includes("eslint-disable")||console.log(r)
```

It looks like a confusion about inlining local functions.

**If the current behavior is a bug, please provide the steps to reproduce.**

See https://github.com/mysticatea/webpack-issue

No loaders, plugins, and configurations.
It just does `webpack --mode production`.

**What is the expected behavior?**

It generates the code which can work.

**Other relevant information:**
webpack version:  4.11.1
Node.js version: 10.4.0
Operating System: Windows 10
Additional tools:

---

This issue was moved from webpack/webpack#7504 by @ooflorent. Original issue was by @mysticatea.

### PR DESCRIPTION
**Pull request by evilebottnawi**
fixes #307 fixes #264

Also we have problem with using option as object, they rewrited after first run.
Also improve readme
A lot of new tests

### CODE REVIEW

File: src/uglify/minify.js [hunk 1]
```diff
@@ -21,17 +21,23 @@ const buildUglifyOptions = ({
...
+  },
   output: {
     shebang: true,
     comments: false,
     beautify: false,
     semicolons: true,
     ...output,
   },
-  // Ignoring sourcemap from options
+  // Ignoring sourceMap from options
```
Comment (evilebottnawi) [inline src/uglify/minify.js]: Typo in option name

File: src/uglify/minify.js [hunk 2]
```diff
@@ -21,17 +21,23 @@ const buildUglifyOptions = ({
...
   ecma,
   warnings,
-  parse,
-  compress,
-  mangle: mangle == null ? true : mangle,
+  parse: {
+    ...parse,
+  },
+  compress: {
+    ...compress,
```
Comment (kzc) [inline src/uglify/minify.js]: this won't handle `compress: false`

File: src/uglify/minify.js [hunk 3]
```diff
@@ -21,17 +21,23 @@ const buildUglifyOptions = ({
...
+  },
+  mangle: mangle == null ? true : {
+    ...mangle,
```
Comment (kzc) [inline src/uglify/minify.js]: this won't handle `mangle: false`

File: src/index.js [hunk 4]
```diff
@@ -42,6 +42,9 @@ class UglifyJsPlugin {
       include,
       exclude,
       uglifyOptions: {
+        compress: {
+          inline: false,
```
Comment (kzc) [inline src/index.js]: `inline: 1` is known to work in `uglify-es@3.3.9`.

```
$ echo '!function(){ console.log(42); }();' | bin/uglifyjs -c inline=false
!function(){console.log(42)}();
```
```
$ echo '!function(){ console.log(42); }();' | bin/uglifyjs -c inline=1
console.log(42);
```

**Comment by kzc** [comment]
If this plugin is upgraded to use `terser` in the future (PR #296), then the `inline` override can be reverted.

File: src/uglify/minify.js (hunk 2 above)
Comment (kzc) [reply src/uglify/minify.js]: nor will it handle `compress: true`

File: src/uglify/minify.js (hunk 3 above)
Comment (kzc) [reply src/uglify/minify.js]: nor will it handle `mangle: true`

**Comment by evilebottnawi** [comment]
@kzc we can not merge use `terser` as `patch/minor` release, it is require major release. 
How we can solve this:
1. Disable `inline` for `1` version (patch release).
2. Allow people using own `cache` keys https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/304 (need implement `function` support) (minor release).
3. Implement feature using own `minify` function (minor release).
4. Documented how plugin can setup using `uglify-js` and `terser` (patch release).
5. Switch on `terser` for next major version (major release).

**Comment by kzc** [comment]
> we can not merge use terser as patch/minor release, it is require major release

I disagree. `terser` is a drop in replacement for `uglify-es`. It's a minor change.

Disabling `inline` in `uglify-es` only addresses a couple of bugs. And it could introduce other bugs in this uglify wrapper plugin as per my PR review comments above.

**Comment by evilebottnawi** [comment]
@kzc other developers disagree with
> I disagree. terser is a drop in replacement for uglify-es. It's a minor change.

All I'm trying to do is find the best solution for everyone and I'm tired of being between these disputes. You suggest using `minor` release (only relying on your words) it's too dangerous and better use major release, it is allow people safely upgrade uglify plugin (If the update introduces new bugs they can roll back without fixing the version in `package.json`).

**Comment by kzc** [comment]
I've been involved with the development of both `uglify-es` and `terser` since their inception. Terser is just a minor fork of `uglify-es` with a name change with dozens of bug fixes. But if you're more comfortable with a major for that, so be it. `webpack@4` users would have to explicitly load the new major plugin to get the bug fixes.

**Comment by evilebottnawi** [comment]
@kzc After release major version with `terser` we wait some times to ensure all works as expected, when bump this in webpack@4

**Comment by shellscape** [comment]
LGTM

**Pull request merged**